storage: exit process when disks are stalled

tbg opened the pull request:
We've had two recent incidents in which we saw clusters with disks
stalled  on a subset of nodes in the cluster. This is a fairly
treacherous failure mode since

- the symptoms are nondescript: from the UI it often looks like a Raft
problem, logspy will freeze up as well, and so you waste some time until
you end up looking at the goroutine dump and notice the writes stuck in
syscall
- the node is in some semi-live state that borders the byzantine and
can cause further trouble for the part of the cluster that isn't
affected (we have some mitigations against this in place but not
enough, and need to improve our defense mechanisms).
- it's sudden and often can't be gleaned from the logs (since everything
is fine and then nothing ever completes so no "alertable" metrics are
emitted).

This commit introduces a simple mechanism that periodically checks for
these conditions (both on the engines and logging) and invokes a fatal
error if necessary.

The accompanying roachtest exercises both a data and a logging disk
stall.

Fixes #7882.
Fixes #32736.

Touches #7646.

Release note (bug fix): CockroachDB will error with a fatal exit when
data or logging partitions become unresponsive. Previously, the process
would remain running, though in an unresponsive state.

knz commented on pkg/server/server_engine_health.go:
```diff
@@ -0,0 +1,83 @@
+	for _, eng := range engines {
+		func() {
+			t := time.AfterFunc(maxDuration, func() {
+				log.Shout(ctx, log.Severity_FATAL, fmt.Sprintf(
```
here and below: this is likely to not-work when crash reporting is enabled. In that case, a FATAL severity will cause things to happen (including disk access) before the call to os.Exit.

So, either:

- ensure the tests for this occur with crash reporting is enabled, or
- modify the code further so that there is a guaranteed call to os.Exit even if the log.Shout never terminates. For example

```go
time.AfterFunc(maxDuration, func ...)
time.Afterfunc(maxDuration + 1 * time.Second, func() { os.Exit(...) })
```

tbg replied:
Ah, I had reviewed `log.Fatal` and saw that it had a case for this and assumed things were handled already. I'll make sure to adopt your suggestion.

petermattis reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> modulo the problem with `LogData` not actually doing anything.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32978#-:-LTOqabtEAdegG6XdcCF:b4ydcvc)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/server/server_engine_health.go, line 41 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/32978#-LTOVyGT1uQV3hhqMBQs:-LTOVyGT1uQV3hhqMBQt:bslmy0d) ([raw file](https://github.com/cockroachdb/cockroach/blob/09f848a3254af10ac6e447c310f9f32fb545bb18/pkg/server/server_engine_health.go#L41)):*
> ```Go
> 				t.Read = true
> 				t.Reset(engineHealthCheckInterval)
> 				assertLoggingHealth(ctx, engineHealthCheckInterval)
> ```

Did you intend to use `engineHealthCheckInterval` for `assertLoggingHealth`? Seems ok, though perhaps that should be renamed as `diskHealthCheckInterval`.
___
*[pkg/server/server_engine_health.go, line 82 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/32978#-LTOW9aZANNXr4kSBZ3R:-LTOW9aZANNXr4kSBZ3S:b-leddd9) ([raw file](https://github.com/cockroachdb/cockroach/blob/09f848a3254af10ac6e447c310f9f32fb545bb18/pkg/server/server_engine_health.go#L82)):*
> ```Go
> 	defer timer.Stop()
> 
> 	log.Flush()
> ```

Doesn't the logging system already have a periodic flush? I believe it just doesn't have the detection of a flush taking an excessively long period of time. 


<!-- Sent from Reviewable.io -->

bdarnell reviewed (approved):
LGTM, but this reminds me of the `server.clock.persist_upper_bound_interval` setting. One of the main reasons we didn't enable that by default (see #25109) is the performance impact of additional disk writes. If we're going to write to disk frequently anyway, maybe we should just add a timeout to it and turn that on by default. (and reduce the number of goroutines performing synced writes to disk).

> Reviewed 6 of 6 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32978#-:-LTTQn9SC8-DlWYpl-Vy:b9nbaz5)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/server/server_engine_health.go, line 29 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/32978#-LTTPUDk7DcPBm75wfk_:-LTTPUDk7DcPBm75wfka:brz4q5f) ([raw file](https://github.com/cockroachdb/cockroach/blob/09f848a3254af10ac6e447c310f9f32fb545bb18/pkg/server/server_engine_health.go#L29)):*
> ```Go
> )
> 
> var engineHealthCheckInterval = envutil.EnvOrDefaultDuration("COCKROACH_ENGINE_HEALTH_CHECK_INTERVAL", 5*time.Second)
> ```

Why an env var instead of a cluster setting? 
___
*[pkg/server/server_engine_health.go, line 82 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/32978#-LTOW9aZANNXr4kSBZ3R:-LTU56TH3FRb4Bphmbaa:bp4kq6i) ([raw file](https://github.com/cockroachdb/cockroach/blob/09f848a3254af10ac6e447c310f9f32fb545bb18/pkg/server/server_engine_health.go#L82)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Doesn't the logging system already have a periodic flush? I believe it just doesn't have the detection of a flush taking an excessively long period of time. 
</blockquote></details>

Yeah, the log package has a goroutine flushing every 30s. This one will be more expensive (or at least more frequent) at 5s.  I'm not sure the cost is a problem, but it doesn't make sense having multiple goroutines performing synchronous log flushes so I'd just add a timeout to the existing log flusher goroutine. 


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
> LGTM, but this reminds me of the server.clock.persist_upper_bound_interval setting. 

Yeah, I can see that, but just jamming this logic in there and making it on by default is more than I have bandwidth for right now. I agree in principle though.

> Dismissed @knz from a discussion.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32978#-:-LTbhiKFBe_nAhgf81BA:b-9nwdos)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/server/server_engine_health.go, line 29 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/32978#-LTTPUDk7DcPBm75wfk_:-LTbh23d-ZSNhIsPDKEq:b2phapy) ([raw file](https://github.com/cockroachdb/cockroach/blob/09f848a3254af10ac6e447c310f9f32fb545bb18/pkg/server/server_engine_health.go#L29)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Why an env var instead of a cluster setting? 
</blockquote></details>

It didn't seem like anything that's reasonable to change. If anything you want an escape hatch for disabling it, which this is. It gets trickier to do a cluster setting now that we're doing something similar in the log package.
___
*[pkg/server/server_engine_health.go, line 41 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/32978#-LTOVyGT1uQV3hhqMBQs:-LTbhBccAuMqlSkzt4L9:brip2xh) ([raw file](https://github.com/cockroachdb/cockroach/blob/09f848a3254af10ac6e447c310f9f32fb545bb18/pkg/server/server_engine_health.go#L41)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Did you intend to use `engineHealthCheckInterval` for `assertLoggingHealth`? Seems ok, though perhaps that should be renamed as `diskHealthCheckInterval`.
</blockquote></details>

Moot now that logging is doing its own thing.
___
*[pkg/server/server_engine_health.go, line 82 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/32978#-LTOW9aZANNXr4kSBZ3R:-LTbhFbiD--lzCwRFAkz:b-q7kbe6) ([raw file](https://github.com/cockroachdb/cockroach/blob/09f848a3254af10ac6e447c310f9f32fb545bb18/pkg/server/server_engine_health.go#L82)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Yeah, the log package has a goroutine flushing every 30s. This one will be more expensive (or at least more frequent) at 5s.  I'm not sure the cost is a problem, but it doesn't make sense having multiple goroutines performing synchronous log flushes so I'd just add a timeout to the existing log flusher goroutine. 
</blockquote></details>

Moved the logging logic to the periodic flusher.


<!-- Sent from Reviewable.io -->

bdarnell reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32978)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/server/server_engine_health.go, line 29 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/32978#-LTTPUDk7DcPBm75wfk_:-LTcCbn92F6U5n6iCWZR:btcsle9) ([raw file](https://github.com/cockroachdb/cockroach/blob/09f848a3254af10ac6e447c310f9f32fb545bb18/pkg/server/server_engine_health.go#L29)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

It didn't seem like anything that's reasonable to change. If anything you want an escape hatch for disabling it, which this is. It gets trickier to do a cluster setting now that we're doing something similar in the log package.
</blockquote></details>

But the log package isn't using this env var now. The log flush loop doesn't really bother me because it becomes a no-op if no log messages have been written (or at least it should), so really it's just deferred I/O from logging and an idle cluster won't do anything there. This one is generating pure new load so there's more reason to want to control it. I'd slightly prefer not to introduce an env var here - either make it a cluster setting or leave it unconfigurable.  


<!-- Sent from Reviewable.io -->

petermattis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32978)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/server/server_engine_health.go, line 29 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/32978#-LTTPUDk7DcPBm75wfk_:-LTcJ4FW8fvzInJyhekp:b8qhjl9) ([raw file](https://github.com/cockroachdb/cockroach/blob/09f848a3254af10ac6e447c310f9f32fb545bb18/pkg/server/server_engine_health.go#L29)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

But the log package isn't using this env var now. The log flush loop doesn't really bother me because it becomes a no-op if no log messages have been written (or at least it should), so really it's just deferred I/O from logging and an idle cluster won't do anything there. This one is generating pure new load so there's more reason to want to control it. I'd slightly prefer not to introduce an env var here - either make it a cluster setting or leave it unconfigurable.  
</blockquote></details>

FYI, a single disk write can be delayed for a significant period of time if the OS is writing a lot of dirty data. I'll point you to the private issue about this. It might be worthwhile tolerating  longer write delays.
___
*[pkg/storage/node_liveness.go, line 779 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/32978#-LTcIcZG4T6iWTK_VAvY:-LTcIcZG4T6iWTK_VAvZ:b-c16x56) ([raw file](https://github.com/cockroachdb/cockroach/blob/09f848a3254af10ac6e447c310f9f32fb545bb18/pkg/storage/node_liveness.go#L779)):*
> ```Go
> 				errors.Wrapf(err, "unable to update node liveness")
> 			}
> 		}
> ```

I'm curious why you removed this. Don't we still want to check that the store can perform a write quickly before updating liveness? Perhaps you should incorporate the engine health check here. That would also get rid of the env variable which Ben dislikes.


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32978)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/server/server_engine_health.go, line 29 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/32978#-LTTPUDk7DcPBm75wfk_:-LThX1gO8ceO6I9Bp9uf:b-ss5iwt) ([raw file](https://github.com/cockroachdb/cockroach/blob/09f848a3254af10ac6e447c310f9f32fb545bb18/pkg/server/server_engine_health.go#L29)):*
> I'd slightly prefer not to introduce an env var here - either make it a cluster setting or leave it unconfigurable.

Ok, I removed it. I don't want to make it a cluster setting since it seems we're going to want to fold this elsewhere in the future. :crossed_fingers: that we never have to disable it.

> It might be worthwhile tolerating longer write delays.

I'm not sure. We handle gray failure poorly. I'm generally wary of inserting process termination into the system, but if a node regularly (or just once) got 10s stalls this would send ripples through the cluster that we would then spend significant resources investigating. At that point I'd rather know (and I had made this an env var to have an escape hatch in case this is *really* wanted or needed at that point in time, though Ben made me remove it). Perhaps logging can take the place of a panic, though it becomes difficult to assure that the log messages make it out. I'd personally prefer to leave this to someone else. I'm already less satisfied with the compromise than with what I had initially, so I'm more than happy to leave further adjustments to you and/or Ben (who may also not be on the same page).
___
*[pkg/storage/node_liveness.go, line 779 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/32978#-LTcIcZG4T6iWTK_VAvY:-LThY3qv5B78xjMNW1gE:b-i46aog) ([raw file](https://github.com/cockroachdb/cockroach/blob/09f848a3254af10ac6e447c310f9f32fb545bb18/pkg/storage/node_liveness.go#L779)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I'm curious why you removed this. Don't we still want to check that the store can perform a write quickly before updating liveness? Perhaps you should incorporate the engine health check here. That would also get rid of the env variable which Ben dislikes.
</blockquote></details>

I removed this because it was obsolete and Ben was concerned about syncing unnecessarily frequently. I conceptually dislike intertwining code like that, but I suppose it could be justified. I think you and Ben should take a second pass at this after it merges if you feel strongly enough, the main contribution here is the roachtest anyway.


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
TFTR! 
PS @petermattis do you want to take this over? You seem to have concerns and I'm looking to get this one over with.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32978#-:-LThYgHb93c_7OQCMbo_:b3o25qn)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)



<!-- Sent from Reviewable.io -->

petermattis reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32978)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/server/server_engine_health.go, line 29 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/32978#-LTTPUDk7DcPBm75wfk_:-LThrT-oCK6nrkTjmJkR:b-d52xhv) ([raw file](https://github.com/cockroachdb/cockroach/blob/09f848a3254af10ac6e447c310f9f32fb545bb18/pkg/server/server_engine_health.go#L29)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

> I'd slightly prefer not to introduce an env var here - either make it a cluster setting or leave it unconfigurable.

Ok, I removed it. I don't want to make it a cluster setting since it seems we're going to want to fold this elsewhere in the future. :crossed_fingers: that we never have to disable it.

> It might be worthwhile tolerating longer write delays.

I'm not sure. We handle gray failure poorly. I'm generally wary of inserting process termination into the system, but if a node regularly (or just once) got 10s stalls this would send ripples through the cluster that we would then spend significant resources investigating. At that point I'd rather know (and I had made this an env var to have an escape hatch in case this is *really* wanted or needed at that point in time, though Ben made me remove it). Perhaps logging can take the place of a panic, though it becomes difficult to assure that the log messages make it out. I'd personally prefer to leave this to someone else. I'm already less satisfied with the compromise than with what I had initially, so I'm more than happy to leave further adjustments to you and/or Ben (who may also not be on the same page).
</blockquote></details>

Ok. Let's go with this for now. We'll know it is in there and can keep an eye out if this check is overly sensitive to external disturbances on the machine.
___
*[pkg/storage/node_liveness.go, line 779 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/32978#-LTcIcZG4T6iWTK_VAvY:-LThrbvWDSq2QOXbU9kD:b3wpbpj) ([raw file](https://github.com/cockroachdb/cockroach/blob/09f848a3254af10ac6e447c310f9f32fb545bb18/pkg/storage/node_liveness.go#L779)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

I removed this because it was obsolete and Ben was concerned about syncing unnecessarily frequently. I conceptually dislike intertwining code like that, but I suppose it could be justified. I think you and Ben should take a second pass at this after it merges if you feel strongly enough, the main contribution here is the roachtest anyway.
</blockquote></details>

Ack. Thinking about this more I think this is fine. Node liveness might thing a node is alive for a few more seconds when it can't actually write to disk, but that should be ok. 


<!-- Sent from Reviewable.io -->

tbg commented:
Minor fixes and made sure the roachtest actually works. I needed to reintroduce env vars because the roachtest needs them (I could futz with charybde but not worth it at the moment).

bors r=petermattis

tbg committed:
storage: exit process when disks are stalled

We've had two recent incidents in which we saw clusters with disks
stalled  on a subset of nodes in the cluster. This is a fairly
treacherous failure mode since

- the symptoms are nondescript: from the UI it often looks like a Raft
problem, logspy will freeze up as well, and so you waste some time until
you end up looking at the goroutine dump and notice the writes stuck in
syscall
- the node is in some semi-live state that borders the byzantine and
can cause further trouble for the part of the cluster that isn't
affected (we have some mitigations against this in place but not
enough, and need to improve our defense mechanisms).
- it's sudden and often can't be gleaned from the logs (since everything
is fine and then nothing ever completes so no "alertable" metrics are
emitted).

This commit introduces a simple mechanism that periodically checks for
these conditions (both on the engines and logging) and invokes a fatal
error if necessary.

The accompanying roachtest exercises both a data and a logging disk
stall.

Fixes #7882.
Fixes #32736.

Touches #7646.

Release note (bug fix): CockroachDB will error with a fatal exit when
data or logging partitions become unresponsive. Previously, the process
would remain running, though in an unresponsive state.

tbg commented:
Got the acceptance tests passing (h/t @knz)

bors r=petermattis
